pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,776 @@
+
+    def _update_trainable_params(self):
+        """Set the trainable parameters"""
+        self._trainable_params = set(self._par_info)
+
+    def _update(self):
+        """Update all internal tape metadata regarding processed operations and observables"""
+        self._graph = None
+        self._update_circuit_info()
+        self._update_par_info()
+        self._update_trainable_params()
+
+    def _expand(self, stop_at=None):
+        """Expand all operations and tapes in the processed queue.
+
+        Args:
+            stop_at (Sequence[str]): Sequence of PennyLane operation or tape names.
+                An operation or tape appearing in this list of names is not expanded.
+        """
+        new_tape = self.__class__()
+        stop_at = stop_at or []
+
+        for op in self.operations:
+            if op.name in stop_at:
+                new_tape._ops.append(op)
+                continue
+
+            t = op if isinstance(op, QuantumTape) else op.expand()
+            new_tape._prep += t._prep
+            new_tape._ops += t._ops
+
+        new_tape._obs = self._obs
+        return new_tape
+
+    def expand(self, depth=1, stop_at=None):
+        """Expand all operations in the processed to a specific depth.
+
+        Args:
+            depth (int): the depth the tape should be expanded
+            stop_at (Sequence[str]): Sequence of PennyLane operation or tape names.
+                An operation or tape appearing in this list of names is not expanded.
+
+        **Example**
+
+        Consider the following nested tape:
+
+        .. code-block:: python
+
+            with QuantumTape() as tape:
+                qml.BasisState(np.array([1, 1]), wires=[0, 'a'])
+
+                with QuantumTape() as tape2:
+                    qml.Rot(0.543, 0.1, 0.4, wires=0)
+
+                qml.CNOT(wires=[0, 'a'])
+                qml.RY(0.2, wires='a')
+                probs(wires=0), probs(wires='a')
+
+        The nested structure is preserved:
+
+        >>> tape.operations
+        [BasisState(array([1, 1]), wires=[0, 'a']),
+         <QuantumTape: params=3, wires=<Wires = [0]>>,
+         CNOT(wires=[0, 'a']),
+         RY(0.2, wires=['a'])]
+
+        Calling ``.expand`` will return a tape with all nested tapes
+        expanded, resulting in a single tape of quantum operations:
+
+        >>> new_tape = tape.expand()
+        >>> new_tape.operations
+        [PauliX(wires=[0]),
+         PauliX(wires=['a']),
+         Rot(0.543, 0.1, 0.4, wires=[0]),
+         CNOT(wires=[0, 'a']),
+         RY(0.2, wires=['a'])]
+        """
+        old_tape = self
+
+        for _ in range(depth):
+            new_tape = old_tape._expand(stop_at=stop_at)
+
+            if len(new_tape.operations) == len(old_tape.operations) and [
+                o1.name == o2.name for o1, o2 in zip(new_tape.operations, old_tape.operations)
+            ]:
+                # expansion has found a fixed point
+                break
```

Comment:
Would it be cleaner if `Operation.expand()` raised an exception if no expansion existed, rather than simply returning itself? Perhaps that would make it easier to detect here

Reply:
I've modified the logic here. `expand` is now recursive, so there is no need to check for fixed points/there are no useless expansions happening.